ddsrecorder_participants/src/cpp/mcap/McapHandler.cpp

```diff
@@ -0,0 +1,460 @@
+            STR_ENTRY << "Error writting in MCAP a message in topic " << topic.m_topic_name
+        );
+    }
+}
+
+void McapHandler::start()
+{
+    if (!paused_)
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Ignoring start command, instance already started.");
+    }
+    else
+    {
+        logInfo(
+            DDSRECORDER_MCAP_HANDLER,
+            "Starting handler.");
+
+        paused_.store(false);
+        stop_event_thread_();
+    }
+}
+
+void McapHandler::pause()
+{
+    if (paused_)
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Ignoring pause command, instance already paused.");
+    }
+    else
+    {
+        logInfo(
+            DDSRECORDER_MCAP_HANDLER,
+            "Pausing handler.");
+
+        dump_data_();
+
+        paused_.store(true);
+        event_thread_ = std::thread(&McapHandler::event_thread_routine_, this);
+    }
+}
+
+void McapHandler::trigger_event()
+{
+    if (!paused_)
+    {
+        logWarning(
+            DDSRECORDER_MCAP_HANDLER,
+            "Ignoring trigger event command, instance is not paused.");
+    }
+    else
+    {
+        logInfo(
+            DDSRECORDER_MCAP_HANDLER,
+            "Triggering event.");
+        {
+            std::lock_guard<std::mutex> lock(event_cv_mutex_);
+            event_triggered_ = true;
+        }
+        event_cv_.notify_one();
+    }
+}
+
+void McapHandler::add_data_(
+        const Message& msg)
+{
+    std::lock_guard<std::mutex> guard(write_mtx_);
+
+    samples_buffer_.push(msg);
+    if (!paused_ && samples_buffer_.size() == buffer_size_)
+    {
+        logInfo(DDSRECORDER_MCAP_HANDLER, "Full buffer, writting to disk...");
+        dump_data_nts_();
+    }
+}
+
+void McapHandler::add_pending_samples_(
+        const std::string& schema_name)
+{
+    assert(pending_samples_.find(schema_name) != pending_samples_.end());
+    auto& pending_queue = pending_samples_[schema_name];
+
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Sending pending samples of type: " << schema_name << ".");
+
+    while (!pending_queue.empty())
+    {
+        auto& sample = pending_queue.front();
+        DdsTopic sample_topic;
+        sample_topic.m_topic_name = sample.first;
+        sample_topic.type_name = schema_name;
+        auto channel_id = get_channel_id_(sample_topic);
+        auto& msg = sample.second;
+        msg.channelId = channel_id;
+        try
+        {
+            add_data_(msg);
+        }
+        catch(const utils::Exception& e)
+        {
+            throw utils::InconsistencyException(
+                STR_ENTRY << "Error writting in MCAP a message in topic " << sample.first
+            );
+        }
+        pending_queue.pop();
+    }
+}
+
+void McapHandler::event_thread_routine_()
+{
+    while (paused_)
+    {
+        clear_all_();
+
+        bool timeout;
+        {
+            std::unique_lock<std::mutex> lock(event_cv_mutex_);
+
+            timeout = !event_cv_.wait_for(
```

Comment:
This is not working as a sliding window, but as a fix division.
e.g. windows size = 20s , handler paused for 30s , event triggered -> store 10s (expected 20s)

Reply:
Refactored to behave as expected.